aten/src/ATen/native/BlasKernel.cpp

```diff
@@ -436,18 +436,32 @@ static C10_ALWAYS_INLINE float32x4_t f32_fma_bf16(float32x4_t a, uint16x4_t b, u
   return f32_fma(a, to_bfloat16(b), to_bfloat16(c));
 }
 
+static C10_ALWAYS_INLINE float32x4_t f32_dot_bf16(float32x4_t a, bfloat16x8_t b, bfloat16x8_t c) {
+#ifdef __ARM_FEATURE_BF16
+  return vbfdotq_f32(a, b, c);
+#else
+  TORCH_CHECK(false, "unsupported operation used!");
+#endif
+}
+
 static C10_ALWAYS_INLINE void dot_with_fp32_arith_main_inner_loop(
   const at::BFloat16* vec1,
   const at::BFloat16* vec2,
   float32x4_t sum[kF32RegistersPerIteration],
   int registerPairIndex) {
-  // TODO: detect intrinsic availability, use them if they're available. __ARM_FEATURE_BF16
+#ifdef __ARM_FEATURE_BF16
+  const bfloat16x8_t temp_vec1 = vld1q_bf16(&vec1[registerPairIndex * 2 * kF32ElementsPerRegister]);
+  const bfloat16x8_t temp_vec2 = vld1q_bf16(&vec2[registerPairIndex * 2 * kF32ElementsPerRegister]);
+  // TODO: we leave half of sum unused. Does this cause suboptimal code generation?
+  sum[registerPairIndex] = f32_dot_bf16(sum[registerPairIndex], temp_vec1, temp_vec2);
```

Comment:
@digantdesai

Reply:
@swolchok [`bfmmla`](https://developer.arm.com/architectures/instruction-sets/intrinsics/vbfmmlaq_f32) is something you can try too. It has 2x compute density.

Reply:
Can you say more about your TODO comment?

Reply:
> @swolchok [bfmmla](https://developer.arm.com/architectures/instruction-sets/intrinsics/vbfmmlaq_f32) is something you can try too. It has 2x compute density.

I will admit to not intuitively understanding why that doesn't do redundant work -- if we want `a dot b` and we stack into `[a1 a2] matmul [b1 b2]^T`, we will get `[[(a1 dot b1) (a1 dot b2)] [(a2 dot a1) (a2 dot b2)]]`. The correct answer is `(a1 dot b1) + (a2 dot b2)`; why do we need the other half of the results?

Reply:
Apologies for a delayed response. I somehow dropped this.

Yeah it will do redundant work. I was alluding something like [this](https://github.com/google/XNNPACK/blob/master/src/qd8-f32-qc4w-gemm/gen/qd8-f32-qc4w-gemm-1x16c8-minmax-neoni8mm.c#L126). BFMMLA is useful for GEMM given its structure, and not so much for GEMV.